### PR TITLE
Add Mixture-of-Miners endpoint

### PR DESCRIPTION
**Pull request by dbobrenko**
## Changes
  - Add Mixture-of-Miners endpoint.
  - Move chat completions processing and mixture of miners into two separate functions.
  - Add system prompt to the inference task.
  
  
 ## Usage example
To use mixture-of-miners, set `mixture` to True:
```
import openai
IP = "..."
PORT = "..."
API_KEY = "..."
prompt = "..."

client = openai.AsyncOpenAI(
    base_url=f"http://{IP}:{PORT}/v1",
    max_retries=0,
    api_key=API_KEY,
)

result = await client.chat.completions.create(
    model=None,
    messages=[{"role": "user", "content": prompt}],
    stream=stream,
    extra_body={
        "seed": seed,
        "sampling_parameters": {
            "temperature": 0.7,
            "top_p": 0.95,
            "top_k": 50,
            "max_new_tokens": 256,
            "do_sample": True,
            "seed": None,
        },
        "task": "QuestionAnsweringTask",
        "mixture": True,
    },
)
```

### CODE REVIEW

File: prompting/api/gpt_endpoints/mixture_of_miners.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import copy
+
+from prompting.api.gpt_endpoints.process_completions import process_completions
+
+DEFAULT_SYSTEM_PROMPT = """You have been provided with a set of responses from various open-source models to the latest user query.
```
Comment (Hollyqui) [inline prompting/api/gpt_endpoints/mixture_of_miners.py:5]: If you have tested this thoroughly ignore this comment: To me putting this into the system prompt feels a little strange. My understanding is that this should simply be appended as a user message, but again I could be wrong on this so feel free to ignore

File: prompting/tasks/task_registry.py [hunk 2]
```diff
@@ -37,42 +28,49 @@ def __hash__(self):
...
-        TaskConfig(
-            task=SummarizationTask, probability=0.1, datasets=[WikiDataset], reward_model=SummarizationRewardConfig
-        ),
-        TaskConfig(
-            task=DateQuestionAnsweringTask,
-            probability=0.1,
-            datasets=[WikiDateDataset],
-            reward_model=DateQARewardConfig,
-        ),
+        # TaskConfig(task=QuestionAnsweringTask, probability=0.2, datasets=[WikiDataset], reward_model=QARewardConfig),
```
Comment (Hollyqui) [inline prompting/tasks/task_registry.py:31]: Please change before merging

File: scripts/test_api.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        async for chunk in result:
+            print(chunk)
+            if chunk.choices[0].delta.content:
+                chunks.append(chunk.choices[0].delta.content)
+        return "".join(chunks)
+
+
+async def main():
+    PORT = 8004
+    API_KEY = "0566dbe21ee33bba9419549716cd6f1f"
```
Comment (Hollyqui) [inline scripts/test_api.py:85]: I think this API key is disposable, please check if that is the case and remove otherwise

**Review by Hollyqui — CHANGES_REQUESTED** [review]
See other comments

File: scripts/test_api.py (hunk 3 above)
Comment (dbobrenko) [reply scripts/test_api.py:85]: I think this is a part of code from your API branch? :)

File: prompting/api/gpt_endpoints/mixture_of_miners.py (hunk 1 above)
Comment (dbobrenko) [reply prompting/api/gpt_endpoints/mixture_of_miners.py:5]: It's based on original paper:
https://arxiv.org/pdf/2406.04692

File: prompting/api/gpt_endpoints/mixture_of_miners.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+Your task is to synthesize these responses into a single, high-quality and concise response.
+It is crucial to follow the provided instuctions or examples in the given prompt if any, and ensure the answer is in correct and expected format.
+Critically evaluate the information provided in these responses, recognizing that some of it may be biased or incorrect.
+Your response should not simply replicate the given answers but should offer a refined and accurate reply to the instruction.
+Ensure your response is well-structured, coherent, and adheres to the highest standards of accuracy and reliability.
+Responses from models:"""
+
+TASK_SYSTEM_PROMPT = {
```
[the remark this answers is not in the document]
Comment (dbobrenko) [reply prompting/api/gpt_endpoints/mixture_of_miners.py:13]: yes, web retrieval would require custom instruction

**Review by bkb2135 — APPROVED** [review]

**Review by richwardle — APPROVED** [review]
LGTM

**Review by Hollyqui — APPROVED** [review]
LGTM

**Pull request merged**